Reorganize the "Canonical Built-ins" section of Explainer.md. (#413)
WebAssembly/component-model: design/mvp/Explainer.md

##### Resource built-ins

###### `resource.new`

| Synopsis                   |                               |
| -------------------------- | ----------------------------- |

> | Conceptual signature       | `func<T>(repr: repr<T>) -> T` |

Reviewer: Thanks!  And for the second point: how about `func<T>(rep: T.rep) -> T` and then say that `T.rep` refers to the `$rep` in the `(type $T (resource (rep $rep) ...)` type definition that defined `T` which is currently hard-coded to `i32`?  I was also thinking that, while perhaps "repr" is a nicer abbreviation, there's already `rep` in WAT and having two slightly-different abbreviations might be confusing and suggest there's two concepts at play.

> | Conceptual signature       | `func<T>(repr: repr<T>) -> T` |

Reviewer: Oh, also, I forgot to mention last time, but, just like `resource.drop T`, `resource.new T` can also only be used in the component that defined `T`, which might be good to include in the abovementioned paragraph, and perhaps synchronize with `resource.drop`, since they're both in the same boat w.r.t `rep`.

> | Conceptual signature       | `func<T>(repr: repr<T>) -> T` |

Author: Sounds good. I've now renamed `repr<T>` to `T.rep`, documented that it refers to the `$rep` and that it's currently hard-coded to `u32`, changed other `repr`s to `rep`s, and added similar language about validation restricting where `resource.drop` and `resource.new` can be used.

also [Returning](Async.md#returning) in the async explainer and
[`canon_task_return`] in the Canonical ABI explainer.)

###### 🔀 `task.wait`

| Synopsis                   |                                           |
| -------------------------- | ----------------------------------------- |

> | Conceptual signature       | `func() -> [kind:event, payload:payload]` |
> | Canonical ABI signature    | `[payload_addr:i32] -> [kind:i32]`        |

Reviewer: ```suggestion
| Conceptual signature       | `func() -> record { kind: event, payload1: u32, payload2: u32 }` |
| Canonical ABI signature    | `[payloads_addr:i32] -> [kind:i32]`        |
```
Pre-existing bug: there are now 2 payload i32s.  Could you update the text below too.

> ```wit

Reviewer: ````suggestion
where `event` is defined in WIT as:
```wit
````

enum event {
  call-starting,
  call-started,
  call-returned,
  call-done,
  yielded,
  stream-read,
  stream-write,
  future-read,
  future-write,
}
```

The `task.wait` built-in returns an event code and
the payload of the event at the address passed as parameter.

In the Canonical ABI, the returned payload is a 4-byte value and is stored
at address `payload_addr`.

`task.wait` can be called whether or not `async` was present, allowing any sort
of code to synchronously wait for progress on any of the currently-executing
subtasks. (See also [Waiting](Async.md#waiting) in the async explainer and
[`canon_task_wait`] in the Canonical ABI explainer.)

###### 🔀 `task.poll`

| Synopsis                   |                                                          |
| -------------------------- | -------------------------------------------------------- |

> | Conceptual signature       | `func() -> [option<tuple<kind:event, payload:payload>>]` |

Reviewer: ```suggestion
| Conceptual signature       | `func() -> option<record { kind: event, payload1: u32, payload2:u32 }> ` |
```

The `subtask.drop` built-in removes the indicated
[subtask](Async.md#subtask-and-supertask) from the current instance's subtask
table, trapping if the subtask isn't done.

###### 🔀 `stream.new` and `future.new`

| Synopsis                              |                          |
| ------------------------------------- | ------------------------ |
| Conceptual signature for `stream.new` | `func<T>() -> stream<T>` |
| Conceptual signature for `future.new` | `func<T>() -> future<T>` |
| Canonical ABI signature               | `[] -> [i32]`            |

The `stream.new` and `future.new` built-ins return the
[writable end](Async.md#streams-and-futures) of a new `stream<T>` or
`future<T>`. (See
[`canon_stream_new`] in the Canonical ABI explainer for details.)

###### 🔀 `stream.read` and `stream.write`

| Synopsis                   |                                                             |
| -------------------------- | ----------------------------------------------------------- |
| Conceptual signature       | `func<T>(stream: stream<T>, buffer: list-buffer<T>) -> stream-status`  |
| Canonical ABI signature    | `[stream:i32 ptr:i32 num:i32] -> [i32]`                     |

> ```wit

Reviewer: ````suggestion
where `stream-status` is defined in WIT as:
```wit
````

enum stream-status {
   // The operation completed and read or wrote this many elements.

>    // This value is always non-zero.

Reviewer: ```suggestion
```
Pre-existing bug, but with `{stream,future}.cancel-{read,write}`, it can now be zero without the stream/future having been closed.

>    complete(u64),

Reviewer: ```suggestion
   complete(u32),
```

Reviewer: I worry this adds some confusion, though, since `u64` is still a bounded-sized integer (as opposed to say `nat`).  When we add a `memory64` `canonopt`, I think we'll need to explicitly either duplicate or parameterize all these tables, in which case we'd update the `complete` case too, so I'd prefer to start with a `u32` that matches for now and explicitly generalize later.

   // The operation did not complete immediately, so callers must wait for
   // the operation to complete by using `task.wait` or by returning to the
   // event loop.

>    blocked,

Reviewer: ```suggestion
   blocked(task),
```

   // The stream is closed. For reading, the end of the stream has been
   // reached. For writing, the reader is no longer reading data.

>    closed,

Reviewer: ```suggestion
   closed(option<error>)
```

Reviewer: Actually, this is only true for the result of `stream.read`; `stream.write`'s `closed` case has no payload.  I suppose that means we should split them up.  This will be nice because then you can use `writable-buffer<T>` and `readable-buffer<T>` (in anticipation of adding them as real WIT types).

}
```

The `future.{read,write}` built-ins
take the matching [readable or writable end](Async.md#streams-and-futures)
of a future as the first parameter, and a buffer for a single `T` value to
read into or write from. The return value is either `complete` if the future
value was eagerly read or written, a sentinel indicating that the
operation did not complete yet (`blocked`), or a sentinel indicating
that the future is closed (`closed`).

In the Canonical ABI, the buffer is passed as a pointer to a buffer
in linear memory. (See
(See [`canon_future_read`] in the Canonical ABI explainer for details.)

###### 🔀 `stream.cancel-read`, `stream.cancel-write`, `future.cancel-read`, and `future.cancel-write`

| Synopsis                                       |                                            |
| ---------------------------------------------- | ------------------------------------------ |
| Conceptual signature for `stream.cancel-read`  | `func<T>(in: stream<T>) -> cancel-status`  |
| Conceptual signature for `stream.cancel-write` | `func<T>(out: stream<T>) -> cancel-status` |
| Conceptual signature for `future.cancel-read`  | `func<T>(in: future<T>) -> cancel-status`  |

> | Conceptual signature for `future.cancel-write` | `func<T>(out: future<T>) -> cancel-status` |

Reviewer: What do you think about using `{readable,writable}-{stream,future}<T>` types in all the built-in descriptions (complementing the `{readable,writable}-buffer` types I suggested above)?

cancellation finished eagerly, the return value is `complete`, and provides
the number of elements read
or written into the given buffer (`0` or `1` for a `future`). If cancellation
blocks, the return value is `blocked` and the caller
must `task.wait`. If the stream or future is closed, the return value is
`closed`.

In the Canonical ABI with the `callback` option, returning to the event
loop is equivalent to a `task.wait`, and a
`{STREAM,FUTURE}_{READ,WRITE}` event will be delivered to indicate the
completion of the `read`
or `write`. (See [`canon_stream_cancel_read`] in the Canonical ABI explainer
for details.)

TODO: Describe how `cancel-status` is mapped to the `i32` return.
TODO: Should `blocked` have a `task`?
TODO: Should `closed` have an `error-context` for the `future.cancel-read` and `stream.cancel-read` cases?

###### 🔀 `stream.close-readable`, `stream.close-writable`, `future.close-readable`, and `future.close-writable`

| Synopsis                                            |                                    |
| --------------------------------------------------- | ---------------------------------- |

> | Approximate WIT signature for `stream.cancel-read`  | `func<T>(in: readable-stream<T>)`  |
> | Approximate WIT signature for `stream.cancel-write` | `func<T>(out: writable-stream<T>)` |
> | Approximate WIT signature for `future.cancel-read`  | `func<T>(in: readable-future<T>)`  |
> | Approximate WIT signature for `future.cancel-write` | `func<T>(out: writable-future<T>)` |

Reviewer 2: Typos & optional error-context on `.close-write`s:

```suggestion
| Approximate WIT signature for `stream.close-read`  | `func<T>(in: readable-stream<T>)`  |
| Approximate WIT signature for `stream.close-write` | `func<T>(out: writable-stream<T>, err: option<error-context>)` |
| Approximate WIT signature for `future.close-read`  | `func<T>(in: readable-future<T>)`  |
| Approximate WIT signature for `future.close-write` | `func<T>(out: writable-future<T>, err: option<error-context>)` |
```

> | Canonical ABI signature                             | `[i32] -> []`                      |

Reviewer 2: `[i32, i32] -> []` for `.close-write`s

Author: Done.

| -------------------------------------------- | ----------------------------------------------------------------------------------- |
| Approximate WIT signature for `future.read`  | `func<T>(in: readable-future<T>, buffer: writable-buffer<T; ..1>) -> read-status`   |
| Approximate WIT signature for `future.write` | `func<T>(out: writable-future<T>, buffer: readable-buffer<T; ..1>) -> write-status` |
| Canonical ABI signature                      | `[future:i32 ptr:i32] -> [i32]`                                                     |

where `read-status` is defined as in [`stream.read`](#-streamread)
and `write-status` is defined as in [`stream.write`](#-streamwrite).

The `future.{read,write}` built-ins
take the matching [readable or writable end](Async.md#streams-and-futures)
of a future as the first parameter, and a buffer for a single `T` value to
read into or write from. The return value is either `complete` if the future
value was eagerly read or written, a sentinel indicating that the
operation did not complete yet (`blocked`), or a sentinel indicating
that the future is closed (`closed`).

The number of elements returned when the value is `complete` is at most `1`.

> The `..1` in the buffer types indicates that these buffers may hold at most
> one element.

Reviewer: ```suggestion
The `<T; 1>` in the buffer types indicates that these buffers may hold at most
one `T` element.
```

| Approximate WIT signature for `stream.cancel-write` | `func<T>(out: writable-stream<T>) -> write-status`  |
| Approximate WIT signature for `future.cancel-read`  | `func<T>(in: readable-future<T>) -> read-status`    |
| Approximate WIT signature for `future.cancel-write` | `func<T>(out: writable-future<T>) -> write-status`  |
| Canonical ABI signature                             | `[i32] -> [i32]`                                    |

where `read-status` is defined as in [`stream.read`](#-streamread)
and `write-status` is defined as in [`stream.write`](#-streamwrite).

The `stream.cancel-read`, `stream.cancel-write`, `future.cancel-read`, and `future.cancel-write`
built-ins
take the matching [readable or writable end](Async.md#streams-and-futures)
of a stream or future that has an outstanding `blocked` read or write. If
cancellation finished eagerly, the return value is `complete`, and provides
the number of elements read
or written into the given buffer (`0` or `1` for a `future`). If cancellation
blocks, the return value is `blocked` and the caller
must `task.wait`. If the stream or future is closed, the return value is
`closed`.

> 
> For `futures.*`, the
> number of elements returned when the value is `complete` is always `1`.

Reviewer: ```suggestion
```
I believe it's possible to return `complete` with `0` if the cancellation eagerly finishes before anything has been read/written from the buffer.

Author: I think that sentence is still useful, so I kept it and just corrected it to say "at most 1", which is similar to what `future.read` and `future.write` say.

Reviewer: Sounds good.  Last nit: s/`futures.*`/`future.*`/

| Approximate WIT signature  | `func(subtask: task)` |
| Canonical ABI signature    | `[i32] -> []`         |

The `subtask.drop` built-in removes the indicated
[subtask](Async.md#subtask-and-supertask) from the current instance's subtask
table, trapping if the subtask isn't done.
(See [`canon_subtask_drop`] in the Canonical ABI explainer for details.)

###### 🔀 `stream.new` and `future.new`

| Synopsis                                   |                          |
| ------------------------------------------ | ------------------------ |
| Approximate WIT signature for `stream.new` | `func<T>() -> stream<T>` |
| Approximate WIT signature for `future.new` | `func<T>() -> future<T>` |
| Canonical ABI signature                    | `[] -> [i32]`            |

The `stream.new` and `future.new` built-ins return the
[writable end](Async.md#streams-and-futures) of a new `stream<T>` or
`future<T>`. (See
[`canon_stream_new`] in the Canonical ABI explainer for details.)

###### 🔀 `stream.read`

| Synopsis                   |                                                             |
| -------------------------- | ----------------------------------------------------------- |

> | Approximate WIT signature  | `func<T>(stream: readable-stream<T>, buffer: readable-buffer<T>) -> read-status`  |

Reviewer: Yeah, good point, n/m!

| Canonical ABI signature    | `[stream:i32 ptr:i32 num:i32] -> [i32]`                     |

where `read-status` is defined in WIT as:
```wit
enum read-status {
   // The operation completed and read this many elements.
   complete(u64),

   // The operation did not complete immediately, so callers must wait for
   // the operation to complete by using `task.wait` or by returning to the
   // event loop.
   blocked(task),

   // The end of the stream has been reached.
   closed(option<error-context>),
}
```

The `stream.read` built-in
takes the matching [readable end](Async.md#streams-and-futures)

> of a stream as the first parameter and a buffer for `T` values to read into.

Reviewer: ```suggestion
of a stream as the first parameter and a buffer for the `T` values read from
the stream to be written into.
```

The return value is either the number of elements (possibly zero) that have
been eagerly read, a sentinel indicating
that the operation did not complete yet (`blocked`), or a sentinel
indicating that the stream is closed (`closed`) along with an
optional error context describing the error that caused to the
stream to close.

In the Canonical ABI, the buffer is passed as a pointer to a buffer
in linear memory and the size in elements of the buffer. (See
[`canon_stream_read`] in the Canonical ABI explainer for details.)

> TODO: Describe how `read-status` is mapped to the `i32` return.

Reviewer: You could perhaps link to the `pack_async_copy_results` function?

###### 🔀 `stream.write`

| Synopsis                   |                                                             |
| -------------------------- | ----------------------------------------------------------- |

> | Approximate WIT signature  | `func<T>(stream: writable-stream<T>, buffer: writable-buffer<T>) -> write-status`  |

Reviewer: ```suggestion
| Approximate WIT signature  | `func<T>(stream: writable-stream<T>, buffer: readable-buffer<T>) -> write-status`  |
```

| Canonical ABI signature    | `[stream:i32 ptr:i32 num:i32] -> [i32]`                     |

where `write-status` is defined in WIT as:
```wit

> enum write-status {

Reviewer: Same comments as `read-status`

   // The operation completed and wrote this many elements.
   complete(u64),

   // The operation did not complete immediately, so callers must wait for
   // the operation to complete by using `task.wait` or by returning to the
   // event loop.
   blocked(task),

   // The reader is no longer reading data.
   closed,
}
```

The `stream.write` built-in
takes the matching [writable end](Async.md#streams-and-futures)

> of a stream as the first parameter and a buffer for `T` values to
> write from.

Reviewer: ```suggestion
of a stream as the first parameter and a buffer from which the `T` values are
to be read by the readable end of the stream.
```

The return value is either the number of elements (possibly zero) that have
been eagerly written, a sentinel indicating
that the operation did not complete yet (`blocked`), or a sentinel
indicating that the stream is closed (`closed`).

In the Canonical ABI, the buffer is passed as a pointer to a buffer
in linear memory and the size in elements of the buffer. (See
[`canon_stream_write`] in the Canonical ABI explainer for details.)

TODO: Describe how `write-status` is mapped to the `i32` return.

###### 🔀 `future.read` and `future.write`

| Synopsis                                     |                                                                   |
| -------------------------------------------- | ----------------------------------------------------------------- |

> | Approximate WIT signature for `future.read`  | `func<T>(in: future<T>, buffer: one-buffer<T>) -> future-status`  |
> | Approximate WIT signature for `future.write` | `func<T>(out: future<T>, buffer: one-buffer<T>) -> future-status` |

Reviewer: ```suggestion
| Approximate WIT signature for `future.read`  | `func<T>(in: readable-future<T>, buffer: writable-buffer<T; ..1>) -> future-status`  |
| Approximate WIT signature for `future.write` | `func<T>(out: writable-future<T>, buffer: readable-buffer<T; ..1>) -> future-status` |
```
and then perhaps comment that the `..1` means that at most one element is read/written from/to the buffer.

| Canonical ABI signature                      | `[future:i32 ptr:i32] -> [i32]`                                   |

where `future-status` is defined in WIT as:
```wit
enum future-status {
   // The operation completed and read or wrote the value.
   complete,

   // The operation did not complete immediately, so callers must wait for
   // the operation to complete by using `task.wait` or by returning to the
   // event loop.
   blocked,

   // The stream is closed. For reading, the end of the stream has been
   // reached. For writing, the reader is no longer reading data.

>    closed,

Reviewer: I think you need the `closed(option<error-context>)` that is only present for `future.read`, as with `stream.read`.

instructions because browsers expect this functionality to come from existing
Web/JS APIs.

###### 🧵 `thread.spawn`

| Synopsis                   |                                                    |
| -------------------------- | -------------------------------------------------- |

> | Approximate WIT signature  | `func<FuncTy, ArgTy>(f: FuncTy, c: ArgTy) -> bool` |
> | Canonical ABI signature    | `[f:(ref null $f) c:i32] -> [i32]`                 |

Reviewer: ```suggestion
| Approximate WIT signature  | `func<FuncT>(f: FuncT, c: FuncT.params[0]) -> bool`       |
| Canonical ABI signature    | `[f:(ref null (func shared (param i32))) c:i32] -> [i32]` |
```

The `thread.spawn` built-in
spawns a new thread by invoking the shared function `f` while passing `c` to it,

> returning whether a thread was successfully spawned.

Reviewer: ```suggestion
returning whether a thread was successfully spawned.  While it's designed to allow
different types in the future, the type of `c` is currently hard-coded to always be `i32`.
```
